reactioncommerce/api-service-account
# Admin UI access control

Signed-off-by: Loan Laux <loan@outgrow.io>

Impact: **critical**
Type: **feature**

This pull request adds a first version of the admin UI access control feature, which includes:

- An `adminUIShopIds` field added to the `Account` schema, which lists the IDs of the shops each user can administer through `reaction-admin`
- A migration which populates the `adminUIShopIds` field for older shops based on `owner` groups in the `Groups` collection
- An `adminUIShops` field added to the `Account` GraphQL schema which returns an array of shops matching the account's `adminUIShopIds`, query-able with the `viewer` query
- `grantAdminUIAccess`, `revokeAdminUIAccess` and `updateAdminUIAccess` mutations to control admin UI shop/account assignments

These were the guidelines provided by @aldeed on [this issue](https://github.com/reactioncommerce/reaction-admin/issues/93#issuecomment-598295994).

## Breaking changes
None.

## Testing
1. Create a shop with an account, confirm that this shop gets added to the account's `adminUIShopIds` array in the `Accounts` collection.
2. Use this PR on a previously-used instance of Reaction Commerce. The migration should be mandatory, and the owner accounts should get their respective shop IDs assigned in the `Accounts` collection as `adminUIShopIds` upon running it.
3. Querying an account (for example with the `viewer` query) should return a list of the shops it has admin UI access to under the `adminUIShops` field.
4. Grant admin UI access to a shop for an account with `grantAdminUIAccess`. Confirm that the shop gets pushed to `adminUIShopIds` in the account's `Accounts` object.
5. Revoke admin UI access from a shop for an account with `revokeAdminUIAccess`. Confirm that the shop gets pulled from `adminUIShopIds` in the account's `Accounts` object.
6. Bulk-update admin UI access to multiple shops for one or multiple accounts with `updateAdminUIAccess`. Confirm that the `adminUIShopIds` get updated in the accounts' `Accounts` objects.
